### PR TITLE
Decouple geode in sadd,srem,smembers

### PR DESCRIPTION
**Pull request by prettyClouds**
This commit builds a GeodeRedisSetSynchronized class that brings
together "set" methods.  These three are necessary for Spring-Session so
they are the only ones included.

Hey team,

The commit brings together two ideas in this PR:

1.  It synchronizes all calls to the set (at least sadd,srem,smembers) via compute.
2. It decouples all geode related code from the Executors.  We have discussed this a couple times, but I wanted to show it, and see if you all are on board.

### CODE REVIEW

File: geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java [hunk 1]
```diff
@@ -306,7 +306,7 @@
    * @return {@link RegionShortcut}
    */
   private static RegionShortcut setRegionType() {
-    String regionType = System.getProperty(DEFAULT_REGION_SYS_PROP_NAME, "PARTITION");
+    String regionType = System.getProperty(DEFAULT_REGION_SYS_PROP_NAME, "REPLICATE");
```
Comment (upthewaterspout) [inline geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java:309]: Why the change to REPLICATE region type as the default? Should it be PARTITION_REDUNDANT, perhaps?

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+          numDeleted.incrementAndGet();
+        }
+      }
+      return newHash;
+    });
+    return numDeleted.intValue();
+  }
+
+  @Override
+  public Collection<Map.Entry<ByteArrayWrapper, ByteArrayWrapper>> hgetall() {
```
Comment (upthewaterspout) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:99]: Because you set methods are never modifying the old value in place, I think would be safe just to return the result of region.get() here, without the compute or the copying.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+  @Override
+  public int hset(List<ByteArrayWrapper> fieldsToSet,
+      boolean NX) {
+    AtomicInteger fieldsAdded = new AtomicInteger();
+
+    Map<ByteArrayWrapper, ByteArrayWrapper> computedHash =
+        region().compute(key, (_unused_, oldHash) -> {
+
+          fieldsAdded.set(0);
```
Comment (upthewaterspout) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:48]: Might want to comment about why you called fieldsAdded.set(0) (because compute can be called multiple times).

The logic you used in the set methods to just call fieldsAdded.set(newHash.size() - oldHash.size()) might be less confusing to someone reading the code.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    if (computedHash != null) {
+      context.getKeyRegistrar().register(this.key, RedisDataType.REDIS_HASH);
+    }
+
+    return fieldsAdded.get();
+  }
+
+  @Override
+  public int hdel(List<ByteArrayWrapper> subList) {
+    AtomicLong numDeleted = new AtomicLong();
```
Comment (upthewaterspout) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:84]: Need to set numDeleted to 0 at the beginning of the compute block.

File: geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java (hunk 1 above)
Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/GeodeRedisServer.java:309]: i was just messing around...i'll rebase this change out of the PR

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java (hunk 2 above)
Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:99]: i was concerned about a CME if somebody iterates over the returned value...i think the copy is necessary but let me know.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java (hunk 3 above)
Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:48]: yea...i only caught that because there was luckily a test that was counting the return values. i can add something.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java (hunk 4 above)
Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:84]: good catch...no test for this, so i missed it.  you are a human junit...DanUnit

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java (hunk 3 above)
Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:48]: i changed the logic...i was hesitant to run .size(), but it's not a concurrenthashmap, so it should be O(1) and consistent

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java (hunk 4 above)
Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:83]: fixed...changed the logic to call AtomicLong.set(<value>) like you recommended

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java (hunk 2 above)
Comment (upthewaterspout) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/hash/GeodeRedisHashSynchronized.java:96]: Well, the function you pass into computeIfPresent is also technically iterating over the oldHash. And since the function is not necessarily executed under lock if anything changes oldHash the function could also through a CME. The compute* methods don't necessarily lock while executing your function, they could be doing a replace followed by a retry of your function for a concurrent modification exception.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/GeodeRedisSetSynchronized.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.geode.cache.Region;
+import org.apache.geode.redis.internal.ByteArrayWrapper;
+import org.apache.geode.redis.internal.ExecutionHandlerContext;
+
+class GeodeRedisSetSynchronized implements RedisSet {
+
+  private ByteArrayWrapper key;
+  private ExecutionHandlerContext context;
+
+  public GeodeRedisSetSynchronized(ByteArrayWrapper key, ExecutionHandlerContext context) {
```
Comment (jhutchison) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/GeodeRedisSetSynchronized.java:32]: I would suggest a refactor to the name here-  this class seems like it's more about operating on sets than a container for data?
